libc/runtime/zipos.internal.h

```diff
@@ -16,7 +16,10 @@ struct ZiposUri {
 };
 
 struct ZiposHandle {
-  struct ZiposHandle *next;
+  union {
+    struct ZiposHandle *next;
+    ssize_t rc;
```

Comment:
@jart should this be atomic?

Reply:
I'm assuming `rc` is a reference count (note: `rc` usually means "result code" in the cosmopolitan codebase). Cosmo has also assumed until now that any given `fd` has a single owner thread. With your change it appears that changes, at least for the `ZiposHandle` data structure. So I'd say, yes we want this to be atomic. We might even need a mutex too (but hopefully not since mutexes aren't a great thing to have in async-signal-safe functions).

Here's how cosmo currently implements reference counting for PosixThread objects.

```c
__funline void _pthread_ref(struct PosixThread *pt) {
  atomic_fetch_add_explicit(&pt->pt_refs, 1, memory_order_relaxed);
}

static bool _pthread_deref(struct PosixThread *pt) {
  int refs = atomic_load_explicit(&pt->pt_refs, memory_order_acquire);
  return !refs || !atomic_fetch_sub(&pt->pt_refs, 1);
}

void _pthread_unref(struct PosixThread *pt) {
  if (_pthread_deref(pt)) {
    unassert(_weaken(_pthread_free));
    _weaken(_pthread_free)(pt, false);
    _weaken(_pthread_decimate)();
  }
}
```

Reply:
Hmm. At least zipos is read-only, so IIUC the only thing that could be contended for besides the refcount is `pos`.

My inclination would be to see if we can get away with making just the two of them atomic. I'm foggy on atomics though, so not sure what memory order makes sense for `pos`.

Reply:
Renamed `rc` to `refs` and made it atomic. Added a TODO for `pos`.

Reply:
I've kept the quirks of `__zipos_keep` / `__zipos_free` for now — different naming convention and using 0 for initial refcount. Can change if desired, or change the pthread implementation (I think 0 for initial ref is slightly nicer.)